package/iframe/main.js

```diff
@@ -257,9 +305,34 @@
     log('Activity support?', !!navigator.mozSetMessageHandler);
     if (navigator.mozSetMessageHandler) {
         navigator.mozSetMessageHandler('activity', function(req) {
-            log('Activity name:', req.source.name);
-            log('Activity data:', JSON.stringify(req.source.data));
-            activitiesToSend.push({name: req.source.name, data: req.source.data});
+            log('Activity name: ', req.source.name);
+            log('Activity data: ', JSON.stringify(req.source.data));
+            var msg = {
+                name: req.source.name,
+                data: req.source.data
+            };
+            if (req.source.name === 'marketplace-openmobile-acl') {
+                // For each activity expecting a returnValue (at the moment
+                // only "marketplace-openmobile-acl", keep the request around,
+                // generating an unique id. When we receive back a message
+                // saying an activity is done, if the id matches one we have,
+                // post the result back to the activity caller).
+                msg.id = getRandomInt(0, Number.MAX_SAFE_INTEGER);
+                activitiesInProgress[msg.id] = req;
+                log('This activity needs to return, generated id: ', msg.id);
+                // 'marketplace-openmobile-acl' also needs to wait on a
+                // getFeature() promise before sending the activity message.
+                if (typeof navigator.getFeature !== 'undefined') {
```

Comment:
Should we call `navigator.hasFeature('navigator.getFeature')` instead?


`</troll>`